R/writeMSData.R

```diff
@@ -0,0 +1,186 @@
+#'     `MSnExp` object and return it in a format that can be passed to the
+#'     `writeMSData` or `copyWriteMSData` functions from the `mzR` package.
+#'
+#' @param x `OnDiskMSnExp` or `MSnExp` object.
+#'
+#' @param software_processing If provided it will be appended to the software
+#'     processing guessed from the object. See [mzR::writeMSData()] for the
+#'     format.
+#' 
+#' @return `list` with the software processing(s).
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+.guessSoftwareProcessing <- function(x, software_processing = NULL) {
+    res <- list()
+    msnbase_proc <- c("MSnbase", paste0(packageVersion("MSnbase"),
+                                        collapse = "."))
+    processings <- processingData(x)@processing
+    if (length(processings)) {
+        proc_cv <- unlist(lapply(processings, FUN = .pattern_to_cv))
+        ## Remove MS:-1 if we have also some mapped processings.
+        if (!all(proc_cv == "MS:-1"))
+            proc_cv <- proc_cv[proc_cv != "MS:-1"]
+        msnbase_proc <- c(msnbase_proc, proc_cv)
+    }
+    res[[1]] <- msnbase_proc
+    if (length(software_processing)) {
+        if (!is.list(software_processing))
+            software_processing <- list(software_processing)
+        res <- c(res, software_processing)
+    }
+    res
+}
+
+
+#' @description `.pattern_to_cv` performs a mapping of pattern to PSI-MS terms.
+#'
+#' @details The mapping bases on a manually curated list of pattern-CV-term
+#'    pairs.
+#' 
+#' @param pattern `character(1)` with the pattern for which a CV term should be
+#'     returned.
+#'
+#' @param ifnotfound `character(1)` to be returned if none of the cv terms
+#'     matches the pattern.
+#' 
+#' @param return `character` with the PSI-MS term or the value of `ifnotfound`
+#'     if it was not found. Note that the length of the character can be > 1 if
+#'     multiple terms match.
+#' 
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+.pattern_to_cv <- function(pattern, ifnotfound = "MS:-1") {
+    res <- ifnotfound
+    if (length(pattern) > 1) {
+        warning("length of pattern is > 1, using only first element")
+        pattern <- pattern[1]
+    }
+    ## character vector mapping pattern to PSI-MS term.
+    .PATTERN.TO.CV <- c(
+        filter = "MS:1001486",
+        normali = "MS:1001484",
+        calibration = "MS:1001485",
+        pick = "MS:1000035",
+        centroid = "MS:1000035",
+        smooth = "MS:1000542",
+        baseline = "MS:1000543",
+        alignment = "MS:1000745"
+    )
+    ## Note: we are matching each names(.PATTERN.TO.CV) to `pattern`, not the
+    ## other way round. So we can provide a string describing the processing
+    ## step and return the best matching CV terms.
+    matches <- vapply(names(.PATTERN.TO.CV), FUN = function(z) {
+        length(grep(pattern = z, x = pattern, ignore.case = TRUE)) > 0
+    }, FUN.VALUE = logical(1))
+    if (any(matches))
```

Comment:
Instead of "preallocating" `res` in the first line. We could simply use
```r
if (any(matches))
    .PATTERN.TO.CV[matches]
else
   ifnotfound
```

Reply:
I'll do.